Updated `harmonized_sdd()` function

juandlt-csu opened the pull request:
I have made the following updates to the harmonize sdd workflow based on my last PR:

- Minor data prep
    - Take absolute values from the original result column
- Remove fails
    - Fail text has been updated to be specific to SDD
    - remove activity and laboratory comment column here.
- Clean up MDLs
    - We just look for "<" to MDL, gets flagged post harmonization
- Clean up "greater than" values
    - We just look for ">" and some other special characters, gets flagged post harmonization
        - Keep the data and remove the equal size, do the same for the stars
    - Have been updated to match those updates in chla
    - The other instances of potentially salvageable data are not considered because they constitute far less than 1% of the total dataset.
- Clean up approximated values/Gap fill harmonized values with depth columns
    - Flagged pre-harmonization
    - There are instances of approximate language, but those are unrelated to SDD. Very few instances of "*"
    - Gap filling harmonized values with depth column information based on depth column criteria
        - flag 0 = value and units not adjusted
        - flag 1 = value filled using activity depth measure column
        - flag 2 = value filled using activity bottom depth measure column
    - Values that have been gap filled will be tier 3.
- Remove remaining NAs
    - If the value or units are empty we drop them. We do not attempt to fill in units with depth column information.
    - In my original PR the idea of retaining bottom values was floating around, but since we are not doing that that is no longer a consideration.
- Harmonize value units
    - In the PR there was an issue with dropping all the values that were NA due to bottom language in the original result column. I have been told that it's okay to drop those values now and to not worry about retaining them and the pipeline reflects that decision.
    - We drop rows where there is no unit to convert.
- Gap fill, harmonize, and flag bottom depth data
    - Flagged pre-harmonization
    - Use this step to gap fill and harmonize bottom depth and ignore the other bits that are in this step for chla.
        - flag 0 = bottom depth value not adjusted and no indication that SD hit bottom
        - flag 1 = bottom depth value adjusted due to indication that SD hit bottom
- Aggregate and tier analytical methods
    - Retain those methods that have "unspecified" in their value. The unrelated text strings have been updated.
    - Using harmonized_local_time to determine time that the observation was taken
    - Sdd_scope_tag has been updated and takes on an approach similar to that of bottom language. There is scope tag and negate scope tag language being considered now.
    - Tiers:
        - 0 = Restrictive. Time reported within the allotted window and method indicates use of viewscope.
        - 1 = Narrowed. Either time reported within the allotted window or method indicates use of viewscope, but not both.
        - 2 = Inclusive. Time not reported/time reported outside of window, and no indication of viewscope used.
        - 3 = Inclusive. harmonized value gap filled with depth column.
- Flag
    - MDLs
        - 0 = value not adjusted and value is greater than 0.01
        - 1 = original NA value adjusted using MDL method
        - 2 = provided value below 0.01; not adjusted
    - Greater than
        - 0 = value less than 31m with NO special characters
        - 1 = value less than 31m with special characters
        - 2 = value greater than 31m with NO special characters
        - 3 = value greater than 31m with special characters
    - Field methods
        - 0 = negate adverse text present (calm day)
        - 1 = neither negate key words nor key words present
        - 2 = key words related to adverse weather conditions present, no negate terms present (windy day)
    - Misc flag
        - 0 = no misc flag
        - 1 = bottom indicated
        - 2 = Harmonized value > bottom value
- Unrealistic values:
    - Not using this step. Values greater than 31m are flagged.
- Generate plots with harmonized dataset
    - Updated to match sdd updates
- Aggregate simultaneous records
    - Updates to simultaneous records have been incorporated
- Export
    - Updated to match sdd updates

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  # comment_cols 
+  comment_cols <- c("ActivityCommentText", "ResultCommentText")
+  # key field text
+  adverse_text <- paste0(c("wind", "chop", "choppy", "precipitation", "rain", "ice"),
```
Additionally, I think the 'wind' text might need to be a bit more specific, this is catching all the comments with wind direction, not actual indications of high/heavy/strong winds.

steeleb replied:
I'd lean towards the 2nd option, and let someone else play with that. If it were more rows with valuable info I'd consider doing more here, but just identifying when weather was reported seems like a good/sane way to deal with this

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+      #       1 = bottom indicated
+      #       2 = Harmonized value > bottom value
+      misc_flag = case_when(
+        bottom_tag == 1 ~ 1,
```
```suggestion
        bottom_tag == 1 & negate_bottom_tag == 0 ~ 1,
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-    arrange(desc(n))
+  # Unrealistic values ------------------------------------------------------
+  
+  # We remove unrealistically high values prior to the final data export
+  
+  # Not applicable to SDD
+  
```
Comment: There are a bunch of super unrealistic values in here (like, greater than 100m)... I know we flag things greater than 30, but I think we also need to filter completely unrealistic values.

juandlt-csu replied:
What should be the cutoff for realistic values? Double 31m?

steeleb replied:
That seems reasonable to me!

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  # First tag aggregate subgroups with group IDs
+  
+  grouped_sdd <- cleaned_flagged_sdd %>%
+    group_by(parameter.x, OrganizationIdentifier, MonitoringLocationIdentifier,
```
I had to edit this to `parameter` to get it to run. :)

juandlt-csu replied:
Updated

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  tier_dists <- no_simul_sdd %>% 
+    select(tier, harmonized_value) %>% 
+    mutate(plot_value = harmonized_value+0.001,
+           tier_label = case_when(
+             tier == 0 ~ "Restrictive (Tier 0)",
+             tier == 1 ~ "Narrowed (Tier 1)",
+             tier == 2 ~ "Inclusive (Tier 2)",
+             tier == 3 ~ "Inclusive (Tier 3)"
+           )) %>% 
```
Suggestion: if you set these up as a factor with levels, you can add a label to it without adding a column AND the fig will be ordered correctly.

```
factor(tier, 
    levels = c(0,1,2,3), 
    labels = c("Restrictive (Tier 0)", "Narrowed (Tier 1)", "Inclusive (Tier 2)", "Inclusive (Tier 3)"))
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    ylab("Count") +
+    ggtitle(label = "Distribution of harmonized SDD CVs by tier",
+            subtitle = "0.001 added to each value for the purposes of visualization only") +
+    scale_x_log10(label = label_scientific()) +
+    scale_y_continuous(label = label_scientific()) +
```
Suggestion: this particular plot seems well-served to not have a log transformation , IMO.
```suggestion
```

juandlt-csu replied:
This makes sense to me

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+             tier == 0 ~ paste0("Restrictive (Tier 0) NAs removed: ", tier_0),
+             tier == 1 ~ paste0("Narrowed (Tier 1) NAs removed: ", tier_1),
+             tier == 2 ~ paste0("Inclusive (Tier 2) NAs removed: ", tier_2),
+             tier == 3 ~ paste0("Inclusive (Tier 2) NAs removed: ", tier_3)
```
```suggestion
             tier == 3 ~ paste0("Inclusive (Tier 3) NAs removed: ", tier_3)
```
Also, see the suggestion about using factors above.

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
   
-  write_feather(grouped_equipment_sdd,
-                data_out_path)
+  write_csv(no_simul_chla,
```
```suggestion
  write_csv(no_simul_sdd,
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
     paste0(
       "Final number of records: ",
-      nrow(grouped_equipment_sdd)
+      nrow(no_simul_chla)
```
```suggestion
      nrow(no_simul_sdd)
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-    harmonized_sdd_path = data_out_path,
-    compiled_drops_path = documented_drops_out_path))  
+    sdd_tiering_record_path = tiering_record_out_path,
+    sdd_grouped_preagg_path = grouped_chla_out_path,
```
```suggestion
    sdd_grouped_preagg_path = grouped_sdd_out_path,
```

steeleb reviewed (changes requested):
Hey Juan!

Great job continuing to work through this!

I left a lot of comments, but most of them are straightforward, I think. LMK if you want to chat at all. I'm going to submit a PR to add the updated config, _targets, and 3d_harmonize_sdd to be able to work within the targets architecture more easily so that no one needs to recreate the wheel there.

mbrousil commented:
@juandlt-csu Sorry, I haven't had a chance to review yet! Should be able look through this early next week.

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -16,21 +12,27 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    # Filter out non-target media types
+    filter(ActivityMediaSubdivisionName %in% c('Surface Water', 'Water', 'Estuary') |
+             is.na(ActivityMediaSubdivisionName)) %>% 
+    # Turn every value into an absolute value
+    mutate(ResultMeasureValue = abs(ResultMeasureValue)) %>% 
```
This is the first and only step that alters data from this column (ResultMeasureValue) before we look for NA's in that column during the MDL step. I can move this out of this step and into the "MDL" or "greater than" step because both of those have to do with altering values that have special characters in them, and I would prefer that to adding them to the approximate flag because the definition of that flag is so different. However, if we do decide to put those values under the approximate flag because of their "-" special character, should we also move those values that have had "=" removed from them under the approximate flag as well? That would mean that the MDL step would only look for "<" special characters and the greater than step would only look for ">" special characters. 

TLDR: We can move this to just before the MDL step, but this is already the only alteration to the ResultMeasureValue column before that step (sans failure removal). We can add an approximate flag to cover these alterations as well, but it might make more sense to put them in the MDL or greater than step instead.

What do y'all think @steeleb  @mbrousil

steeleb replied:
I think functionally, it makes sense to deal with them before MDL/greater than (but in that same processing procedure. Essentially, we're altering the values, which could add uncertainty (say, that character was not supposed to be a '-' but a '9' or something), which we should document. In terms of the approximate flag for MDL/greater than, I don't think it's necessary because they are already flagged as such.

juandlt-csu replied:
```
sdd_absvalues <- sdd_fails_removed %>%
    # Filter for negative sign
    filter(grepl("-", ResultMeasureValue))

sdd_absvalues <- sdd_fails_removed %>%
    # Filter for negative numbers (integers or floats)
    filter(grepl("^-\\d+(\\.\\d+)?$", ResultMeasureValue))
```

Both of these objects return 187 rows, do we want to add a flag for this?

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-  # Replace harmonized_value field with these new values
-  sdd_harmonized_values <- sdd_approx_added %>%
+  # Replace "harmonized_value" field with these new values
+  sdd_harmonized_values <- sdd_mdls_added %>%
```
I think this naming convention followed chla when is was trying to convert it over, but I think this change makes sense.

juandlt-csu replied:
made this update

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -42,139 +44,124 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-  sdd_fails_removed <- sdd %>%
+  # The values that will be considered fails for each column:
+  fail_text <- c(
+    "error", "fail", "invalid", "no result",  "questionable", "suspect", 
+    "unable", "reject", "no data", "Not Reported", "no reading"
```
Made these updates

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -42,139 +44,124 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  )
+  
+  # Now get counts of fail-related string detections for each column: 
+  fail_counts <- list("ResultCommentText", "ResultMeasureValue_original") %>%
```
```
table(sdd %>% 
    filter(grepl(paste(fail_text, collapse = "|"), ResultDetectionConditionText)) %>% 
    pull(ResultDetectionConditionText))
```

This results in 286 entries for the fail text "Not Reported" with no other fail text hits. Most of these do not carry any values in the `ResultMeasureValue_original` (i.e. contain NA or #VALUE!). Given how inconsequential dropping these would be I will add `ResultDetectionConditionText` to this step.

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    # Plot and unit codes that didn't make through joining
+    sdd_no_na %>%
+      anti_join(x = .,
+                y = unit_conversion_table,
```
I think that is correct

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    # prefer to fill with activity column if bottom language is not indicated or there is negate language
+    # can only gap fill when both value and units are present for the depth columns we are going to use
+    mutate(
+      # flag 0 = value and units not adjusted
+      # flag 1 = value filled using activity depth measure column
+      # flag 2 = value filled using activity bottom depth measure column
```
Going to include the negative approximation in the misc flags

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+        depth_flag == 1 ~ "bottom depth value adjusted due to indication that SD hit bottom.",
+        .default = NA_character_
+      ),
+      harmonized_bottom_depth_value = as.numeric(harmonized_bottom_depth_value) * conversion,
+      harmonized_bottom_depth_unit = if_else(!is.na(harmonized_bottom_depth_unit), "m", NA),
```
We will re-join the conversion table based on harmonized_bottom_depth_unit to resolve this, thanks B!

juandlt-csu commented:
I have made the following updates to the harmonize SDD workflow based on this PR’s comments.
* Minor data prep:
    * Removed taking the absolute value from this step, and instead added a `Clean up special characters` step before the `Clean up MDLs` step.
* Clean up special characters:
    * Look for `-`, `=`, and `*` in the original value column and extract the number values. 
    * This step flags this data under miscellaneous flag 3 = Special characters present (*, =,-)
* Clean up “greater than” values
    * Remove `=` from this string search. Now only searches for `>`
* Clean up approximated values/Gap fill harmonized values with depth columns
    * Simplified the logic for the flagging criteria.
        * Only fill harmonized value if there is Activity depth data or bottom depth data.
        * Prefer to fill in with activity depth data
        * Only prefer bottom depth data if there is bottom language and no negate language
* Gap fill, harmonize, and flag bottom depth data
    * Re-joined the conversion table based on harmonized bottom depth for bottom depth gap fill unit conversion. Previously we had used the conversion column from the harmonized units conversion step under the assumption that these units would be the same.
* Flag
    * Converted adverse text to environmental conditions text. Field flag is now used to determine if there was environmental information stored in the comment columns instead of determining if there were good/bad weather conditions. 
* Unrealistic values:
    * Cutting off those values that are greater than 62m (double our cutoff for flagging suspiciously high values)
* General:
    * Made those updates that deal with downstream effects of these changes

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
   print(
     paste0(
-      "Rows removed due to fails, missing data, etc.: ",
+      "Rows removed due to fail-related language: ",
```
quibble: I wonder if this should be percentage for consistency?

juandlt-csu replied:
Updated to percent

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-  sdd_approx <- sdd_approx %>%
-    select(approx_value, index)
-  
+
```
comment: add `rm()` & `gc()` up here, too

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+      harmonized_units = ResultMeasure.MeasureUnitCode,
+      harmonized_comments = "Approximated during 'greater than' step")
+
   greater_vals$greater_value <- as.numeric(
```
```suggestion
  # this will cause coercion errors, which is expected, as some values have additional characters in
  # them that cause this step to have errors.
  greater_vals$greater_value <- as.numeric(
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    number_dropped = nrow(sdd_approx_added) - nrow(sdd_greaterthan_added),
+    n_rows = nrow(sdd_approx_added),
+    order = 6
+  )
```
```suggestion
  )
  
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-    count(ResultMeasure.MeasureUnitCode) %>%
+  rm(sdd_greaterthan_added)
+  gc()
+  # Remove remaining NAs ----------------------------------------------------
```
```suggestion

  # Remove remaining NAs ----------------------------------------------------
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  )
+
+  # Free up memory
+  rm(sdd_greaterthan_added, sdd_approx_added, sdd_mdls_added,
+     sdd_fails_removed)
```
```suggestion
  rm(sdd_approx_added)
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  print(
+    paste(
+      round((nrow(sdd_bottom_depth)) / nrow(converted_units_sdd) * 100, 3),
+      "% of samples have bottom depths that can be back filled with harmonized values."
```
```suggestion
      "% of samples have no reported ActivityBottomDepthHeightMeasure.MeasureValue that can be 
      back filled with harmonized values where Secchi on bottom was indicated."
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
     paste0(
-      "Rows removed due to unlikely sample methods: ",
-      nrow(grouped_fractions_sdd) - nrow(grouped_sample_methods_sdd)
+      "Rows removed while assigning flags: ",
```
Flagging: n row or percent

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  # How many records removed due method harmonization?
   print(
     paste0(
       "Rows removed due to unlikely analytical methods: ",
-      nrow(converted_units_sdd) - nrow(grouped_analytical_methods_sdd)
+      nrow(sdd_bottom_depth_added) - nrow(sdd_relevant)
     )
   )
```
Flagging: n row or percent

juandlt-csu replied:
updated to percent

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  # How many records removed in aggregating simultaneous records?
   print(
     paste0(
-      "Rows removed due to unlikely sample methods: ",
-      nrow(grouped_sample_methods_sdd) - nrow(grouped_equipment_sdd)
+      "Rows removed while aggregating simultaneous records: ",
+      nrow(cleaned_flagged_sdd) - nrow(no_simul_sdd)
     )
   )
```
Flagging: n row or percent

juandlt-csu replied:
Updated to percent

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
   # How many records removed due to unit harmonization?
   print(
     paste0(
-      "Rows removed while converting to numeric values: ",
-      nrow(sdd_harmonized_values) - nrow(converted_units_sdd)
+      "Rows removed while converting units: ",
+      nrow(sdd_no_na) - nrow(converted_units_sdd)
     )
   )
```
Flagging: n row or percent

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+      (time_tag != 0) & (scope_tag != 0) ~ 2,
+      .default = 2
+      ),
+      tier = if_else(approx_flag != 0, 3, tier)
```
Just need to adjust from new approx_flag 1 where we use abs for negative values
```suggestion
      tier = if_else(approx_flag > 1, 3, tier)
```

juandlt-csu replied:
approx_flag 1 is still when the value is approximated from `ActivityDepthHeightMeasure.MeasureValue` column. The negative value conversion now falls under the misc flag. Do you want me to update this so that it a record is tier 3 if it is flagged with the misc flag indicating a special character removal?

steeleb replied:
Superb. You can leave as you wrote it then. 

``` 
          tier = if_else(approx_flag != 0, 3, tier)
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  realistic_sdd <- flagged_sdd %>%
+    filter(harmonized_value <= 62)
+
```
suggestion: pop a percent dropped here for consistency

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  # Plot harmonized measurements by CharacteristicName
+
+  plotting_subset <- cleaned_flagged_sdd %>%
```
```suggestion
  plotting_subset <- realistic_sdd %>%
```

juandlt-csu replied:
Updated

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    theme_bw() +
+    theme(strip.text = element_text(size = 7))
+
+  # ggsave(filename = "../AquaMatch_harmonize_WQP_old/scratch/figures/sdd_charname_dists.png",
```
```suggestion
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,988 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-  
-  documented_drops_out_path <- "3_harmonize/out/harmonize_sdd_dropped_metadata.csv"
-  
+  compiled_dropped <- bind_rows(starting_data, dropped_media,
+                                dropped_fails, dropped_specialcharacters, dropped_mdls,
+                                dropped_greater_than, dropped_approximates,
+                                dropped_na, dropped_harmonization,
+                                dropped_depths, dropped_methods,
+                                dropped_field, dropped_simul)
```
Looks like 11 & 13 are missing, 9 is duplicated.
![Screenshot 2024-08-14 at 10 38 54 AM](https://github.com/user-attachments/assets/1322988c-410c-40ac-bf49-c765c59d701d)

juandlt-csu replied:
Updated the tibbles and compiled dropped

steeleb reviewed (changes requested):
So stinking close!! 

Just a few more tweaks. 

Additionally, can we round all the harmonized values that were converted to 2 decimal point precision to avoid false precision due to the conversion? I think this is as easy as `if_else(harmonized_units != ResultMeasureValue.Units, round(harmonized_value, 2), harmonized value)`

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+            (is.na(ActivityBottomDepthHeightMeasure.MeasureUnitCode) & ActivityDepthHeightMeasure.MeasureUnitCode == harmonized_units) |
+            ActivityDepthHeightMeasure.MeasureUnitCode == harmonized_units |
+            ActivityBottomDepthHeightMeasure.MeasureUnitCode == harmonized_units)
+        )
```
Resolving this and related comments as they were fixed when we simplified the approximation code.

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+      # INCLUDE harmonized_value is NA AND...
+      is.na(harmonized_value) &
+        # either ActivityDepthHeightMeasure OR... 
+        ((!is.na(ActivityDepthHeightMeasure.MeasureValue) & !is.na(ActivityDepthHeightMeasure.MeasureUnitCode)) | 
```
Made this update to the simplified approximation logic as well.

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+        # either ActivityDepthHeightMeasure OR... 
+        ((!is.na(ActivityDepthHeightMeasure.MeasureValue) & !is.na(ActivityDepthHeightMeasure.MeasureUnitCode)) | 
+        # ActivityBottomDepthHeightMeasure (and relevant unit columns) are not NA AND...
+        (!is.na(ActivityBottomDepthHeightMeasure.MeasureValue) & !is.na(ActivityBottomDepthHeightMeasure.MeasureUnitCode))) &
```
Updated

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -184,312 +171,887 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+         negate_bottom_tag == 1) & 
+          !is.na(ActivityDepthHeightMeasure.MeasureValue) &
+          (is.na(ResultMeasure.MeasureUnitCode) | # should this be harmonized_units?
+             ResultMeasure.MeasureUnitCode == ActivityDepthHeightMeasure.MeasureUnitCode) ~ 1,
```
Resolved per approximation logic simplification

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  )
+
+  # Now get counts of fail-related string detections for each column:
+  fail_counts <- list("ResultCommentText", "ResultMeasureValue_original",
+                      "ResultDetectionConditionText") %>%
```
question: Should `ActivityCommentText` be included here too? When I added it in there were a decent number of results for fail language.

juandlt-csu replied:
`ActivityCommentText` was removed because a big portion of those rows that were removed did not pertain to SDD.

mbrousil replied:
```suggestion
  # Note that ActivityCommentText not included because fail comments were mostly about non-SDD data
  fail_counts <- list("ResultCommentText", "ResultMeasureValue_original",
                      "ResultDetectionConditionText") %>%
```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-      ) | is.na(ResultMeasureValue_original))
-  
-  # How many records removed due to fails, missing data, etc.?
+      if_all(.cols = c(ResultCommentText, ResultMeasureValue_original),
```
question: Should the same columns used on L53 above be included here?

juandlt-csu replied:
Yes, updated.

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  greater_vals <- sdd_mdls_added %>%
+    # First, remove the samples that we've already approximated:
+    filter((!index %in% mdl_updates$index)) %>%
+    # only want NAs in the altered value column and where there is a `<` and a number in the original values column...
```
```suggestion
    # only want NAs in the altered value column and where there is a `>` and a number in the original values column...
```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    # just in case.
+    mutate(harmonized_value = abs(as.numeric(str_extract(ResultMeasureValue_original, "-?\\d+(\\.\\d+)?"))),
+      harmonized_units = ResultMeasure.MeasureUnitCode,
+      harmonized_comments = "Approximated during 'greater than' step")
```
suggestion: We don't have a `harmonized_comments` column in the chlorophyll data, so it would be ideal to make it so that it's not necessary here. Lmk if this was something I missed in the last meeting though

juandlt-csu replied:
Hm, I forget where I got that from, maybe an old sdd script or the old chla script. I can just remove that column from this workflow, would that work?

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-    select(approx_value, index)
-  
+
+  # Clean up MDLs --------------------------------------------------------------
```
suggestion: This section does a good job of working within the constraints of the previous harmonization script! I do think it would be clearer to readers/users if we didn't describe it as an "MDL" process. I'm flexible on this, but ultimately it feels like a square peg, round hole kind of situation to me. "Less than" is fine with me, like the "greater than" handling later on.

juandlt-csu replied:
how does this fit with the naming conventions in the final dataset output, which has `mdl_flag` as an output.

steeleb replied:
An alternative is that all of these could get lumped in with special characters.

mbrousil replied:
I think that's a great idea @steeleb! It's probably the most conceptually consistent way to organize it too.

juandlt-csu replied:
So we are going to remove the MDL flag altogether, and those values that are less that 0.01 will not get flagged?

steeleb replied:
I think there are two things happening here: 

1) we remove the special character '<' - which in some other parameters indicates BDL/MDL. We don't treat these data as MDL(I don't think...) but rather just remove the special character (which would be a reason to include '<' in special characters alongside '=' and '*'). We *could* add the MDL flag of 2 (MDL indicated, but value unchanged) in the instances where '<' is in the ResultMeasureValue and leave the code for removing that character as-is.

2) we also flag anything < 0.01 m as MDL (a flag of 2 currently, but maybe it should be 1, value reported <0.01, changed to 0.01) after unit conversion because that is the documented physical measure that we think is possible (a human can't discern less than 1cm resolution). 

There are two options here:
1) move "<" into the special character handling with "=" and "*" and not notate in MDL. This would leave MDL virtually blank because the 0.01 flagging happens so much later. Note that we may want to consider altering the harmonized value of anything <0.01 to 0.01. In this case, a MDL flag of 1 is not used and should probably be removed. 

2) Re-think this and change our flags/motivation so that "<" means MDL and it is flagged as such and incorporate these changes into the code. 

So here is what I think we should do:
      # Flag: 0 = value not adjusted and value is greater than 0.01m
      #       1 = original value <0.01m adjusted to 0.01m
      #       2 = original value contained "<" in ResultMeasureValue column, which was removed, but the value was unaltered

mbrousil replied:
Thanks for the explanation! I agree with your suggestion @steeleb.

@juandlt-csu is it possible to do the `mdl_flag` assignments in the MDL section rather than in the Flag section? If not I think it would at least be helpful to describe the MDL situation B has outlined in that section.

juandlt-csu replied:
We flag MDL during that step because it is after we harmonize the units, so I think it's best to keep the MDL flagging in the flagging section . Do we want go with option 1 or 2? I think option 2 makes more sense, since it's more in line with what we have so far.

steeleb replied:
I agree, @juandlt-csu! Let's go for # 2 here - thanks for being flexible as we continually change the rules on you. :)

mbrousil replied:
Yep, agreed on option 2!

juandlt-csu replied:
Great, this will be just like a reverse `greater than` flag then.

juandlt-csu replied:
I've updated the `MDL` flag to be more in line with the `greater than` flag. The conversion of the sub-0.01m values to 0.01m happens in the `Realistic range of values` section (formerly `unrealistic values`), where I added a very short blurb about it and a print statement to know the percentage of rows that were converted.

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  rm(sdd_mdls_added)
   gc()
+
+  # Clean up approximated values/Gap fill harmonized values --------------------
```
suggestion: My personal preference here would be to describe the things happening in L271:370 as "gap filling" rather than approximation

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+        .default = harmonized_value
+      ),
+      harmonized_units = case_when(
+        approx_flag == 2 ~ ActivityBottomDepthHeightMeasure.MeasureValue,
```
Not sure, but I think this needs to be changed?
```suggestion
        approx_flag == 2 ~ ActivityBottomDepthHeightMeasure.MeasureUnitCode,
```

juandlt-csu replied:
Great catch!

steeleb replied:
oops, this needs to be committed too!

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  print(
+    paste(
+      round((nrow(sdd_bottom_depth)) / nrow(converted_units_sdd) * 100, 3),
+      "% of samples have no reported ActivityBottomDepthHeightMeasure.MeasureValue that can be 
+      back filled with harmonized values where Secchi on bottom was indicated."
```
Suggestion: Could this print statement be clarified a little? I'm not entirely sure I understand the meaning. Perhaps:
```suggestion
      "% of samples have no reported ActivityBottomDepthHeightMeasure.MeasureValue.  These will be 
      back filled with harmonized values where Secchi on bottom was indicated."
```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-  # Now count the fraction column (the ungrouped version): 
-  fraction_counts <- grouped_analytical_methods_sdd %>%
-    count(ResultSampleFractionText) %>%
+
+  # time_tag
```
suggestion: Can you include a short comment here outlining why time is being tagged?

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  # time_tag
+  sdd_time_tag <- sdd_relevant %>%
+    mutate(
+      harmonized_local_time = as.POSIXct(harmonized_local_time,
```
question: Can you use the `harmonized_utc` column for this? It's already in UTC. If not, I'd avoid writing over the original `harmonized_local_time` column here. Better to make a temporary column that gets removed instead.

juandlt-csu replied:
I will make a temporary column instead of overwriting the current harmonized_local_time column.

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+      ),
+      harmonized_local_time = format(harmonized_local_time, "%Y-%m-%d %H:%M:%S") # return harmonized_local_time to characters
+    )
+
+  # scope_tag
```
suggestion: Can you include a short comment here outlining why scope text is being tagged?

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  # combinations of the tags that already exist
+
+  # comment_cols
+  comment_cols <- c("ActivityCommentText", "ResultCommentText")
```
issue: Where possible we should avoid reusing variable names because it's possible for someone to accidentally filter the wrong part of the dataset, e.g. if they're hopping around the script at all. Could you rename instances of `comment_cols` so that they're step-specific?

juandlt-csu replied:
Updated

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+    )
+
+  # tier for SDD
+  tiered_methods_sdd <- sdd_scope_tag %>%
```
question: The output of this step contains instances where `tier == 1` and `time_tag == 2`. Is this expected?

juandlt-csu replied:
This is expected if `scope_tag == 0` in those cases, as it would fulfill the `exclusive or` condition.

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  tiering_record_out_path <- "3_harmonize/out/sdd_tiering_record.csv"
+
+  # Slim the tiered product (post flagging)
```
issue: `sdd_tiering_record.csv` doesn't get exported here. This also caused an error when I ran the pipeline because `p3_sdd_tiering_record` is still expected but fails to run.

juandlt-csu replied:
Updated

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  gc()
+
+  # Flag --------------------------------------------------------------------
+  # We are going to flag everything that hasn't been flagged yet during this step
+  # we are not going to worry about bottom language in this step because it is
+  # already in the bottom tag. if people care about these they can make their own
+  # combinations of the tags that already exist
```
suggestion: Could you clarify this comment? I wasn't sure what it meant, I think in particular "flag everything that hasn't been flagged yet"

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+                                      "rain", "calm", "clear"),
+                                    collapse = "|")
+
+  flagged_sdd <- tiered_methods_sdd %>%
```
suggestion: It feels to me like this section makes the `harmonized_comment` column redundant. That could be an easy way to offload the information stored there?

juandlt-csu replied:
Updated to offload harmonized_Comment here

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  flagged_sdd <- tiered_methods_sdd %>%
+    mutate(
+      # MDL
+      # Flag: 0 = value not adjusted and value is greater than 0.01
+      #       1 = original NA value adjusted using MDL method
+      #       2 = harmonized value below 0.01; not adjusted
+      mdl_flag = case_when(
+        (harmonized_value > 0.01) ~ 0,
+        (harmonized_value <= 0.01) ~ 2,
```
question: What's the significance of 0.01 here? Sorry if I missed something!

juandlt-csu replied:
Values less than 0.01m are below our method detection limit. Flag 2 if the SDD reading was reported as less than 1 cm beneath the surface of the water.

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  gc()
+
+  # Unrealistic values ------------------------------------------------------
+
```
suggestion: Could you include a short comment here indicating the source for 62m?

mbrousil reviewed (changes requested):
Thanks for your patience, Juan! I left some comments and questions that need to be addressed. I think they all should be pretty straightforward but let me know if you have any questions about them.

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
     paste0(
-      "Rows removed while converting to numeric values: ",
-      nrow(sdd_harmonized_values) - nrow(converted_units_sdd)
+      "Percentage of rows removed while converting units: ", #***
+      nrow(sdd_no_na) - nrow(converted_units_sdd)
```
```suggestion
      "Percentage of rows removed while converting units: ", #***
      round(((nrow(sdd_no_na)) - nrow(converted_units_sdd)/nrow(sdd_no_na)) * 100, 3)
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  print(
+    paste0(
+      "Percent of rows removed due to unlikely analytical methods: ",
+      round((nrow(sdd_bottom_depth_added)-nrow(sdd_relevant))/nrow(sdd_bottom_depth_added)*100,3)
```
```suggestion
      round((nrow(sdd_bottom_depth_added)-nrow(sdd_relevant))/nrow(sdd_bottom_depth_added)*100,3), 
      " %"
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-      "Rows removed due to unlikely fraction type: ",
-      nrow(grouped_analytical_methods_sdd) - nrow(grouped_fractions_sdd)
+      "Percentage of rows removed while assigning flags: ",
+      round((nrow(tiered_methods_sdd) - nrow(cleaned_flagged_sdd)) / nrow(tiered_methods_sdd) * 100, 3)
```
```suggestion
      round((nrow(tiered_methods_sdd) - nrow(cleaned_flagged_sdd)) / nrow(tiered_methods_sdd) * 100, 3),
      " %"
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  # Percentage of records removed due method harmonization?
+  print(
+    paste0(
+      "Percent of rows removed due to unlikely analytical methods: ",
```
```suggestion
      "Percentage of rows removed due to unlikely analytical methods: ",
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-      "Rows removed due to unlikely sample methods: ",
-      nrow(grouped_fractions_sdd) - nrow(grouped_sample_methods_sdd)
+      "Percentage of rows removed while removing unrealistic values: ",
+      round((nrow(cleaned_flagged_sdd) - nrow(realistic_sdd)) / nrow(cleaned_flagged_sdd) * 100, 3)
```
```suggestion
      round((nrow(cleaned_flagged_sdd) - nrow(realistic_sdd)) / nrow(cleaned_flagged_sdd) * 100, 3),
      " %"
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  print(
+    paste(
+      round((nrow(sdd_no_na)) / nrow(sdd_approx_added) * 100, 2),
```
```suggestion
      round((nrow(sdd_no_na)) / nrow(sdd_approx_added) * 100, 3),
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,997 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-      "Rows removed due to unlikely sample methods: ",
-      nrow(grouped_sample_methods_sdd) - nrow(grouped_equipment_sdd)
+      "Percentage of rows removed while aggregating simultaneous records: ",
+      round((nrow(realistic_sdd) - nrow(no_simul_sdd)) / nrow(realistic_sdd) * 100, 3)
```
```suggestion
      round((nrow(realistic_sdd) - nrow(no_simul_sdd)) / nrow(realistic_sdd) * 100, 3),
      " %"
```

steeleb reviewed (commented):
Just a couple of tiny tweaks for consistency to go along with MB's review. I was able to run this without issue, but didn't dive deeply in!

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -473,7 +473,7 @@ harmonize_sdd <- function(raw_sdd, p_codes){
     paste(
       round((nrow(sdd_bottom_depth)) / nrow(converted_units_sdd) * 100, 3),
-      "% of samples have no reported ActivityBottomDepthHeightMeasure.MeasureValue that can be 
+      "% of samples have no reported ActivityBottomDepthHeightMeasure.MeasureValu. These will be 
```
```suggestion
      "% of samples have no reported ActivityBottomDepthHeightMeasure.MeasureValue but Secchi on bottom has been indicated and a harmonized value is present. These will be 
```
There are additional conditions to describe the % of samples here. Added more context to add readabillity.

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -770,6 +782,15 @@ harmonize_sdd <- function(raw_sdd, p_codes){
+  # This threshold is established based on the following rationale:
+  # 1. SDD values greater than 31 meters are flagged as potentially anomalous in our quality control process.
+  # 2. The 62-meter threshold represents a doubling of this initial flagging criterion.
+  # 3. Measurements beyond this point are considered improbable and likely erroneous.
```
```suggestion
  # 3. Measurements beyond this point are considered improbable and likely erroneous due to unit conversion or other data entry issues.
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -104,161 +104,180 @@ harmonize_sdd <- function(raw_sdd, p_codes){
-    # the result is positive.
-    mutate(harmonized_value = abs(as.numeric(str_extract(ResultMeasureValue_original, "-?\\d+(\\.\\d+)?"))),           # Preserve the units if they are provided 
+    # Extract the first numeric value after '<' from ResultMeasureValue_original and
+    # convert it to a number. This captures values like "<0.5", ignoring any text after the number.
```
@juandlt-csu can you just clarify this one - this is just removing the <, right?

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1032 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  # Clean up MDLs --------------------------------------------------------------
+
+  # Find MDLs and make them usable as numeric data
```
```suggestion
  # Find MDL character ("<") and make them usable as numeric data
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1032 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-  
+
+  # Next step, incorporating "greater than" values. Using a similar approach to
+  # gap filling the MDL ("less than") values, we can gap fill the results that
```
```suggestion
  #  the MDL ("less than") values, we can coerce the ResultMeasureValue that
```

steeleb commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1032 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  # Next step, incorporating "greater than" values. Using a similar approach to
+  # gap filling the MDL ("less than") values, we can gap fill the results that
+  # contain values greater than some amount.
```
```suggestion
  # contain a greater than symbol.
```

steeleb reviewed (commented):
This looks phenomenal.

I have a handful of suggestions, most of which I admit are absolute nit picks, but important nonetheless. I think this can be merged after you address the handful of comments and after MB has re-reviewed.

B

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1032 @@ harmonize_sdd <- function(raw_sdd, p_codes,
     str_replace_all(
       greater_vals$ResultMeasureValue_original,
-      c("\\>" = "", "\\*" = "", "\\=" = "" )))
+      c("\\>" = "", "\\=" = "" )))
```
Correct, updated

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1032 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+                                     harmonized_value, ResultMeasureValue),
+           harmonized_units = ifelse((index %in% mdl_updates$index),
+                                     harmonized_units, ResultMeasure.MeasureUnitCode),
+           harmonized_comments = ifelse(index %in% mdl_updates$index,
```
This column gets offloaded in the `flag` section

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1032 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+                                     ResultMeasure.MeasureUnitCode, harmonized_units),
            harmonized_comments = ifelse(index %in% greater_vals$index,
-                                        "Value identified as being greater than listed value.",
+                                        "Approximated during 'greater than' step",
```
Do we want to change this to "cleaned" instead?

juandlt-csu replied:
Changed this language to cleaned

juandlt-csu commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1032 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+            grepl(">", ResultMeasureValue_original) &
+             grepl("[0-9]", ResultMeasureValue_original)) %>%
+    # Extract the first numeric value after '>' from ResultMeasureValue_original and
+    # convert it to a number. This captures values like ">0.5", ignoring any text after 
+    # the number, and ensures the result is positive.
```
Updated. Solution here was to exclude samples with alphabetical characters from the filter before we extract the number.

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1035 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+
+  # Clean up MDLs --------------------------------------------------------------
+
+  # Find MDL character ("<") and make them usable as numeric data
```
```suggestion
  # Find "less than" ("<") characters and make them usable as numeric data.
  # For other parameters finding "<" more directly relates to method detection
  # limits, but for consistency with flagging columns across datasets we use
  # the same naming conventions for SDD.

```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1035 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  )
+
+
+  # Aggregate and tier analytical methods -----------------------------------
```
```suggestion
# Aggregate analytical methods --------------------------------------------
```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1035 @@ harmonize_sdd <- function(raw_sdd, p_codes,
   dropped_methods <- tibble(
     step = "sdd harmonization",
     reason = "Dropped rows while aggregating analytical methods",
     short_reason = "Analytical methods",
```
```suggestion
    short_reason = "Aggregate methods",
```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1035 @@ harmonize_sdd <- function(raw_sdd, p_codes,
-  # Now count the fraction column (the ungrouped version): 
-  fraction_counts <- grouped_analytical_methods_sdd %>%
-    count(ResultSampleFractionText) %>%
+
```
```suggestion


 # Tier analytical methods -------------------------------------------------

```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1035 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+  dropped_tiers <- tibble(
+    step = "sdd harmonization",
+    reason = "Dropped rows while tiering analytical methods",
+    short_reason = "Analytical methods",
```
```suggestion
    short_reason = "Tier methods",
```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1035 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+      datum = unique(datum)
+    ) %>%
+    # Calculate coefficient of variation as the standardard deviation divided by
+    # the mena value (harmonized_value in this case)
```
```suggestion
    # the mean value (harmonized_value in this case)
```

mbrousil commented on 3_harmonize/src/harmonize_sdd.R:
```diff
@@ -13,483 +9,1035 @@ harmonize_sdd <- function(raw_sdd, p_codes,
+      lat = unique(lat),
+      datum = unique(datum)
+    ) %>%
+    # Calculate coefficient of variation as the standardard deviation divided by
```
```suggestion
    # Calculate coefficient of variation as the standard deviation divided by
```

mbrousil reviewed (approved):
Thanks Juan! This looks good. Just a couple of small edits from me and then it can be merged in.
